solr/core/src/java/org/apache/solr/cluster/placement/SolrCollection.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement;
+
+import java.util.Map;
+
+/**
+ * Represents a Collection in SolrCloud (unrelated to {@link java.util.Collection} that uses the nicer name).
+ */
+public interface SolrCollection {
+  /**
+   * The collection name (value passed to {@link Cluster#getCollection(String)}).
+   */
+  String getName();
+
+  /**
+   * <p>The {@link Shard}'s over which the data of this {@link SolrCollection} is distributed.
+   *
+   * <p>The map is from {@link Shard#getShardName()} to {@link Shard} instance.
+   */
+  Map<String, Shard> getShards();
```

Comment:
Please note that in some extreme (but real!) deployments the number of shards may be in the order of thousands and the number of replicas in the order of hundreds of thousands... Not so sure about a simple Map here - maybe a list of shard names + getShard accessor?

(Or use a read-through pseudo-map like the one added with SOLR-14680 ?)

Reply:
What about replacing this by `Shard getShard(String name);` and `Iterator<Shard> getShardIterator();` ?
These can be implemented efficiently.

Same thing should then be done for `Cluster.getAllCollectionNames()` that I've added reluctantly, maybe returning an `Iterator` is better?

Reply:
If it's an Iterator then I think it should iterate on names only, or maybe have two iterators, one for names and one for Shard-s? I'm beginning to like the `SimpleMap` model from @noblepaul because it gives you a getter + iterators for names and/or values in one abstraction (the name of that class is confusing though ...).

Reply:
What use case is served by an iterator that only iterates on shard names and that doesn't "materialize" them into `Shard` instances?
If it's for finding a `Shard` whose name matches some pattern, we should devise more efficient ways.
I'm not against having such an iterator but need help understanding when it might be useful (given an iterator that returns `Shard` instances and a method to fetch a given `Shard` by name).